fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/StandardIndexMaintainer.java

```diff
@@ -400,6 +423,29 @@ public <M extends Message> List<IndexEntry> filteredIndexEntries(@Nullable final
         return indexEntries;
     }
 
+    @Nullable
+    private <M extends Message> List<IndexEntry> filteredIndexEntriesByPredicate(@Nonnull final FDBIndexableRecord<M> savedRecord) {
+        // Note: for now, IndexPredicate will not support filtering of certain index entries
+        final IndexPredicate predicate = state.index.getPredicate();
+        // Already checked: saved record and predicate are not null,
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug(KeyValueLogMessage.of("index predicate used",
+                    LogMessageKeys.INDEX_NAME, state.index.getName(),
+                    LogMessageKeys.FILTER, predicate));
+        }
+        CorrelationIdentifier objectQuantifier = Quantifier.current();
+        final RecordType recordType = state.store.getRecordMetaData().getRecordType(savedRecord.getRecordType().getName());
+        Type.Record typeRecord = Type.Record.fromDescriptor(recordType.getDescriptor());
+        Value recordValue = QuantifiedObjectValue.of(objectQuantifier, typeRecord);
+        QueryPredicate queryPredicate = predicate.toPredicate(recordValue);
```

Comment:
This looks expensive, is it possible to cache the generated predicate (`queryPredicate`) and only (re)generate it if the incoming `savedRecord` has a different `typeRecord`?

Reply:
I like this idea!

Reply:
From what I'm seeing, the standard maintainer is regenerate by `FDBRecordStore` for every record update. I wonder if the cache should be saved at the index predicate..

Reply:
> I wonder if the cache should be saved at the index predicate..

I think this is fine, although I rather put it in a separate small (caching) factory class so we can keep the POJO contract of the `IndexPredicate` class.